File: android/app/src/main/kotlin/com/knovik/edgetal/EmbedderChannel.kt
package com.knovik.edgetal

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.text.textembedder.TextEmbedder
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import java.util.concurrent.Executors
import kotlin.math.sqrt

/**
 * Native side of the `edgetal/embedder` channel — on-device text embeddings via
 * MediaPipe's Text Embedder.
 */
class EmbedderChannel(private val context: Context) {

    private var embedder: TextEmbedder? = null
    private var dimension: Int = 0
    private val executor = Executors.newSingleThreadExecutor()
    private val main = Handler(Looper.getMainLooper())

    fun register(messenger: BinaryMessenger) {
        MethodChannel(messenger, "edgetal/embedder").setMethodCallHandler { call, result ->
            when (call.method) {
                "initialize" -> executor.execute {
                    try {
                        val dim = ensureEmbedder()
                        Log.i("EmbedderChannel", "Initialization succeeded with dimension $dim")
                        main.post { result.success(dim) }
                    } catch (e: Exception) {
                        Log.e("EmbedderChannel", "Initialization failed", e)
                        main.post {
                            result.error(
                                "INIT_FAILED",
                                "${e.javaClass.simpleName}: ${e.message}",
                                e.stackTraceToString()
                            )
                        }
                    }
                }
                "embed" -> {
                    val text = call.argument<String>("text") ?: ""
                    executor.execute {
                        try {
                            ensureEmbedder()
                            val raw = embedder!!.embed(text)
                                .embeddingResult()
                                .embeddings()
                                .first()
                                .floatEmbedding()
                            val normalized = normalize(raw).map { it.toDouble() }
                            main.post { result.success(normalized) }
                        } catch (e: Exception) {
                            Log.e("EmbedderChannel", "Embedding failed for text length ${text.length}", e)
                            main.post { result.error("EMBED_FAILED", e.message, null) }
                        }
                    }
                }
                else -> result.notImplemented()
            }
        }
    }

    /** Lazily builds the embedder and returns its output dimension. */
    @Synchronized
    private fun ensureEmbedder(): Int {
        if (embedder == null) {
            val options = TextEmbedder.TextEmbedderOptions.builder()
                .setBaseOptions(
                    BaseOptions.builder()
                        .setModelAssetPath("text_embedder.tflite")
                        .build()
                )
                .build()
            embedder = TextEmbedder.createFromOptions(context, options)
            val probe = embedder!!.embed("dimension probe")
            dimension = probe.embeddingResult()
                .embeddings()
                .first()
                .floatEmbedding()
                .size
            Log.i("EmbedderChannel", "MediaPipe Text Embedder probe complete. Dimension=$dimension")
        }
        return dimension
    }

    private fun normalize(v: FloatArray): FloatArray {
        var mag = 0.0
        for (x in v) mag += (x * x).toDouble()
        mag = sqrt(mag)
        if (mag == 0.0) return v
        return FloatArray(v.size) { (v[it] / mag).toFloat() }
    }
}

File: ios/Runner/EmbedderChannel.swift
import Flutter
import Foundation
import MediaPipeTasksText

/// Native side of the `edgetal/embedder` channel on iOS — on-device text embeddings
/// via MediaPipe's Text Embedder (`TextEmbedder`).
class EmbedderChannel {
    private var embedder: TextEmbedder?
    private var dimension: Int = 0
    private let queue = DispatchQueue(label: "com.knovik.edgetal.embedder", qos: .userInitiated)

    func register(messenger: FlutterBinaryMessenger) {
        let channel = FlutterMethodChannel(name: "edgetal/embedder", binaryMessenger: messenger)
        channel.setMethodCallHandler { [weak self] call, result in
            guard let self = self else { return }
            switch call.method {
            case "initialize":
                self.queue.async {
                    do {
                        let dim = try self.ensureEmbedder()
                        DispatchQueue.main.async { result(dim) }
                    } catch {
                        DispatchQueue.main.async {
                            result(FlutterError(code: "INIT_FAILED", message: error.localizedDescription, details: nil))
                        }
                    }
                }
            case "embed":
                let args = call.arguments as? [String: Any]
                let text = args?["text"] as? String ?? ""
                self.queue.async {
                    do {
                        _ = try self.ensureEmbedder()
                        guard let embedder = self.embedder else {
                            throw NSError(domain: "EmbedderError", code: -1, userInfo: [NSLocalizedDescriptionKey: "Embedder instance is nil"])
                        }
                        let embeddingResult = try embedder.embed(text: text)
                        guard let firstEmbedding = embeddingResult.embeddingResult.embeddings.first else {
                            throw NSError(domain: "EmbedderError", code: -2, userInfo: [NSLocalizedDescriptionKey: "No embeddings returned"])
                        }

                        let floatVector: [Float]
                        if let floatArray = firstEmbedding.floatEmbedding {
                            floatVector = floatArray.map { $0.floatValue }
                        } else {
                            floatVector = []
                        }
                        let normalized = self.normalize(floatVector).map { Double($0) }
                        DispatchQueue.main.async { result(normalized) }
                    } catch {
                        DispatchQueue.main.async {
                            result(FlutterError(code: "EMBED_FAILED", message: error.localizedDescription, details: nil))
                        }
                    }
                }
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }

    private func ensureEmbedder() throws -> Int {
        if embedder != nil {
            return dimension
        }
        guard let modelPath = Bundle.main.path(forResource: "text_embedder", ofType: "tflite") else {
            throw NSError(domain: "EmbedderError", code: 404, userInfo: [NSLocalizedDescriptionKey: "text_embedder.tflite asset not found in main bundle"])
        }
        let options = TextEmbedderOptions()
        options.baseOptions.modelAssetPath = modelPath

        let createdEmbedder = try TextEmbedder(options: options)
        self.embedder = createdEmbedder

        let probeResult = try createdEmbedder.embed(text: "dimension probe")
        if let firstEmbedding = probeResult.embeddingResult.embeddings.first,
           let floatArray = firstEmbedding.floatEmbedding {
            dimension = floatArray.count
        } else {
            dimension = 512
        }
        return dimension
    }

    private func normalize(_ v: [Float]) -> [Float] {
        var mag: Double = 0.0
        for x in v { mag += Double(x * x) }
        mag = sqrt(mag)
        if mag == 0.0 { return v }
        return v.map { Float(Double($0) / mag) }
    }
}
